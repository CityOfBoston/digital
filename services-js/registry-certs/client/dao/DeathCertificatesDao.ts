import DataLoader from 'dataloader';
import { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import lookupDeathCertificateOrder from '../queries/lookup-death-certificate-order';
import { DeathCertificateSearchResults, DeathCertificateOrder } from '../types';
import { FetchGraphql } from '@cityofboston/next-client-common';

export type DeathCertificateCache = { [id: string]: DeathCertificate };

export default class DeathCertificatesDao {
  fetchGraphql: FetchGraphql;
  loader: DataLoader<string, DeathCertificate | null>;

  cacheFullQuery: string | null = null;
  cacheByPage: { [page: number]: DeathCertificateSearchResults } = {};

  constructor(fetchGraphql: FetchGraphql) {
    this.fetchGraphql = fetchGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader(ids =>
      fetchDeathCertificates(fetchGraphql, ids).then(a => [...a])
    );
  }

  get(id: string): Promise<DeathCertificate | null> {
    return this.loader.load(id);
  }

  async search(
    fullQuery: string,
    page: number
  ): Promise<DeathCertificateSearchResults> {
    if (fullQuery === this.cacheFullQuery) {
      const results = this.cacheByPage[page];
      if (results) {
        return results;
      }
    } else {
      this.cacheFullQuery = fullQuery;
      this.cacheByPage = {};
    }

    const { query, startYear, endYear } = this.parseQuery(fullQuery);

    const results = await searchDeathCertificates(
      this.fetchGraphql,
      query,
      page,
      startYear,
      endYear
    );

    this.cacheByPage[page] = results;

    results.results.forEach(cert => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }

  async lookupOrder(
    id: string,
    contactEmail: string
  ): Promise<DeathCertificateOrder | null> {
    return lookupDeathCertificateOrder(this.fetchGraphql, id, contactEmail);
  }

  parseQuery(
    fullQuery: string
  ): { query: string; startYear: string | null; endYear: string | null } {
    // match a 4-digit year, and optionally a second 4-digit year with a hypen or en-dash
    const yearRegexp = /(\d{4})\s*[-â€“]?\s*(\d{4})?/;

    let query;
    let startYear;
    let endYear;

    const match = fullQuery.match(yearRegexp);
    if (match) {
      query = fullQuery.replace(yearRegexp, '').trim();
      startYear = match[1];
      endYear = match[2] || match[1];
    } else {
      query = fullQuery;
      startYear = null;
      endYear = null;
    }

    return {
      query,
      startYear,
      endYear,
    };
  }
}
