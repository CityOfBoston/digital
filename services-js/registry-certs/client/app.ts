/* global Stripe */

import { configure as mobxConfigure } from 'mobx';
import Router from 'next/router';
import { hydrate } from 'emotion';

import { Context as NextContext } from 'next';

import RouterListener from './lib/RouterListener';
import SiteAnalytics from './lib/SiteAnalytics';

import makeLoopbackGraphql, { LoopbackGraphql } from './lib/loopback-graphql';

import { RequestAdditions } from '../server/lib/request-additions';

import Cart from './store/Cart';
import OrderProvider from './store/OrderProvider';
import Accessibility from './store/Accessibility';

import DeathCertificatesDao from './dao/DeathCertificatesDao';
import CheckoutDao from './dao/CheckoutDao';

export type ClientContext = NextContext<RequestAdditions>;

export type ClientDependencies = {
  stripe: stripe.Stripe | null;
  loopbackGraphql: LoopbackGraphql;
  cart: Cart;
  accessibility: Accessibility;
  orderProvider: OrderProvider;
  deathCertificatesDao: DeathCertificatesDao;
  checkoutDao: CheckoutDao;
  siteAnalytics: SiteAnalytics;
};

let browserInited = false;
let browserDependencies: ClientDependencies;

const accessibility = new Accessibility();
const siteAnalytics = new SiteAnalytics();

// Browser-only setup
export function initBrowser() {
  if (browserInited) {
    return;
  }

  browserInited = true;

  // Adds server generated styles to emotion cache.
  // '__NEXT_DATA__.ids' is set in '_document.js'
  if (typeof window !== 'undefined') {
    hydrate((window as any).__NEXT_DATA__.ids);
  }

  mobxConfigure({ enforceActions: true });

  accessibility.attach();
  siteAnalytics.attach((window as any).ga);

  const routerListener = new RouterListener();
  routerListener.attach(Router, accessibility, (window as any).ga);
}

// Works on both server and browser. Memoizes on browser, so these dependencies
// will share state across pages.
//
// Typically called from a wrapper function or controller component to inject
// these into a component's props.
export function getDependencies(ctx?: ClientContext): ClientDependencies {
  if ((process as any).browser && browserDependencies) {
    return browserDependencies;
  }

  // req will exist only when this function is called for getInitialProps on the
  // server.
  const req = ctx && ctx.req;

  const stripe =
    typeof Stripe !== 'undefined'
      ? Stripe(
          ((window as any).__NEXT_DATA__ || {}).stripePublishableKey ||
            'test-publishable-key'
        )
      : null;
  const loopbackGraphql = makeLoopbackGraphql(req);
  const deathCertificatesDao = new DeathCertificatesDao(loopbackGraphql);
  const checkoutDao = new CheckoutDao(loopbackGraphql, stripe);
  const cart = new Cart();
  const orderProvider = new OrderProvider();

  if ((process as any).browser) {
    cart.attach(window.localStorage, deathCertificatesDao, siteAnalytics);
    orderProvider.attach(window.localStorage);
  }

  const dependencies: ClientDependencies = {
    stripe,
    accessibility,
    cart,
    orderProvider,
    deathCertificatesDao,
    checkoutDao,
    loopbackGraphql,
    siteAnalytics,
  };

  if ((process as any).browser) {
    browserDependencies = dependencies;
  }

  return dependencies;
}
