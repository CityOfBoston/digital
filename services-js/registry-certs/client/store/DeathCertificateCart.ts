import { observable, computed, action, autorun } from 'mobx';

import { GaSiteAnalytics } from '@cityofboston/next-client-common';

import { DeathCertificate } from '../types';
import DeathCertificatesDao from '../dao/DeathCertificatesDao';

import { CERTIFICATE_COST } from '../../lib/costs';

const DEATH_CERTIFICATE_COST = CERTIFICATE_COST.DEATH;

interface LocalStorageEntry {
  id: string;
  quantity: number;
}

export class DeathCertificateCartEntry {
  id: string = '';
  @observable.ref cert: DeathCertificate | null = null;
  @observable quantity: number = 0;
}

export default class DeathCertificateCart {
  @observable entries: Array<DeathCertificateCartEntry> = [];
  @observable pendingFetches: number = 0;

  localStorageDisposer: Function | null = null;
  siteAnalytics: GaSiteAnalytics | null = null;

  @action
  attach(
    localStorage: Storage | null,
    deathCertificatesDao: DeathCertificatesDao,
    siteAnalytics: GaSiteAnalytics
  ) {
    this.siteAnalytics = siteAnalytics;

    if (localStorage) {
      try {
        const savedCart: Array<LocalStorageEntry> = JSON.parse(
          localStorage.getItem('cart') || '[]'
        );

        this.entries = savedCart
          .filter(({ quantity }) => quantity > 0)
          .map(
            action(
              'hydrate entry from local storage start',
              ({ id, quantity }: LocalStorageEntry) => {
                const entry = new DeathCertificateCartEntry();
                entry.id = id;
                entry.cert = null;
                entry.quantity = quantity;

                this.pendingFetches += 1;

                deathCertificatesDao.get(id).then(
                  action(
                    'hydrate item from local storage complete',
                    (cert: DeathCertificate | null) => {
                      if (cert) {
                        entry.cert = cert;
                      } else {
                        this.remove(id);
                      }
                      this.pendingFetches -= 1;
                    }
                  )
                );

                return entry;
              }
            )
          );
      } catch (e) {
        localStorage.removeItem('cart');
      }

      this.localStorageDisposer = autorun(
        () => {
          localStorage.setItem(
            'cart',
            JSON.stringify(
              this.entries.map(
                ({ id, quantity }): LocalStorageEntry => ({
                  id,
                  quantity,
                })
              )
            )
          );
        },
        {
          name: 'save cart to local storage',
        }
      );
    }
  }

  @action
  detach() {
    if (this.localStorageDisposer) {
      this.localStorageDisposer();
      this.localStorageDisposer = null;
    }

    this.siteAnalytics = null;
  }

  trackCartItems() {
    const { siteAnalytics } = this;

    if (!siteAnalytics) {
      return;
    }

    this.entries.forEach(({ id, quantity }) => {
      siteAnalytics.addProduct(
        id,
        'Death certificate',
        'Death certificate',
        quantity,
        DEATH_CERTIFICATE_COST / 100
      );
    });
  }

  @computed
  get size(): number {
    // quantity shouldn't be below 0 but we want to be defensive.
    return this.entries.reduce(
      (acc, item) => acc + Math.max(item.quantity, 0),
      0
    );
  }

  @computed
  get loading(): boolean {
    return this.pendingFetches > 0;
  }

  setQuantity(cert: DeathCertificate, quantity: number) {
    const { siteAnalytics } = this;

    const existingItem = this.entries.find(({ id }) => id === cert.id);
    const filteredQuantity = Math.max(0, Math.min(99, quantity));

    const quantityChange = quantity - this.getQuantity(cert.id);

    if (siteAnalytics) {
      siteAnalytics.addProduct(
        cert.id,
        'Death certificate',
        'Death certificate',
        Math.abs(quantityChange),
        DEATH_CERTIFICATE_COST / 100
      );
      siteAnalytics.setProductAction(quantityChange < 0 ? 'remove' : 'add');
    }

    if (existingItem) {
      // We don't remove items here when their quantity is 0 so that they don't
      // disappear when you edit the values on the cart page.
      existingItem.quantity = filteredQuantity;
    } else {
      const item = new DeathCertificateCartEntry();
      item.id = cert.id;
      item.cert = cert;
      item.quantity = filteredQuantity;

      this.entries.push(item);
    }
  }

  remove(certId: string) {
    const { siteAnalytics } = this;
    const idx = this.entries.findIndex(({ id }) => id === certId);

    if (idx !== -1) {
      if (siteAnalytics) {
        siteAnalytics.addProduct(
          certId,
          'Death certificate',
          'Death certificate',
          this.getQuantity(certId),
          DEATH_CERTIFICATE_COST / 100
        );
        siteAnalytics.setProductAction('remove');
      }

      this.entries.splice(idx, 1);
    }
  }

  getQuantity(certId: string): number {
    const entry = this.entries.find(({ id }) => id === certId);

    if (entry) {
      return entry.quantity;
    } else {
      return 0;
    }
  }

  clean() {
    this.entries = this.entries.filter(({ quantity }) => quantity > 0);
  }

  @action
  clear() {
    this.entries = [];
  }

  @computed
  get containsPending(): boolean {
    return !!this.entries.find(({ cert }) => !!(cert && cert.pending));
  }
}
