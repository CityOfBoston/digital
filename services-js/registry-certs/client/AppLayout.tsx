/* eslint react/display-name: 0 */

import React, { KeyboardEvent, MouseEvent } from 'react';
import Link from 'next/link';
import { css } from 'emotion';

const headerHtml: string = require('../templates/header.html');
const navigationHtml: string = require('../templates/navigation.html');
const footerHtml: string = require('../templates/footer.html');

import Nav from './common/Nav';
import DeathCertificateCart from './store/DeathCertificateCart';

import { FREEDOM_RED } from './common/style-constants';

// These props only require a Cart if showNav is true. Discriminated unions FTW.
type Props =
  | {
      showNav?: false;
    }
  | {
      showNav: true;
      cart: DeathCertificateCart;
    };

const LAST_BREADCRUMB_STYLE = css({
  color: FREEDOM_RED,
});

const AppLayout: React.StatelessComponent<Props> = props => (
  <div>
    <a
      href="#content-start"
      tabIndex={1}
      className="btn a11y--h a11y--f"
      style={{ margin: 4 }}
      onClick={(ev: MouseEvent<HTMLAnchorElement>) => ev.currentTarget.blur()}
    >
      Jump to content
    </a>

    <input
      type="checkbox"
      id="brg-tr"
      className="brg-tr"
      aria-label="Open Boston.gov menu"
      onKeyDown={(ev: KeyboardEvent<HTMLInputElement>) => {
        if (ev.key === 'Tab' && !ev.shiftKey && !ev.currentTarget.checked) {
          const a = document.querySelector('header a') as HTMLAnchorElement;
          a && a.focus();
          ev.preventDefault();
        }
      }}
    />
    <nav
      className="nv-m"
      dangerouslySetInnerHTML={{ __html: navigationHtml }}
      aria-label="Boston.gov menu"
    />

    <div className="mn--full-ie">
      <div
        className={`mn mn--full ${props.showNav ? 'mn--nv-s' : ''}`}
        style={{ zIndex: 2 }}
      >
        <input type="checkbox" id="s-tr" className="s-tr" aria-hidden="true" />
        <header
          className="h"
          role="banner"
          dangerouslySetInnerHTML={{ __html: headerHtml }}
        />

        {props.showNav && <Nav cart={props.cart} />}

        <nav className="brc p-a300" aria-label="Breadcrumbs">
          <ul className="brc-l">
            <li className="brc-l-i">
              <a href="https://www.boston.gov/">Home</a>
              <span className="brc-s"> › </span>
            </li>
            <li className="brc-l-i">
              <a href="https://www.boston.gov/departments">Departments</a>
              <span className="brc-s"> › </span>
            </li>
            <li className="brc-l-i">
              <a href="https://www.boston.gov/departments/registry">
                Registry: Birth, death, and marriage
              </a>
              <span className="brc-s"> › </span>
            </li>
            <li className="brc-l-i">
              <Link href="/death">
                <a className={LAST_BREADCRUMB_STYLE}>Death certificates</a>
              </Link>
            </li>
          </ul>
        </nav>

        <div id="content-start" className="a11y--content-start" />
        <main className="b-ff">{props.children}</main>
      </div>
    </div>

    <footer
      className="ft"
      style={{ position: 'relative', zIndex: 2 }}
      dangerouslySetInnerHTML={{ __html: footerHtml }}
    />
  </div>
);
export default AppLayout;
