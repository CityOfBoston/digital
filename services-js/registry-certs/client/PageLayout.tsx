/* eslint react/display-name: 0 */

import React from 'react';

import { AppLayout, BreadcrumbNav } from '@cityofboston/react-fleet';

import DeathCertificateCart from './store/DeathCertificateCart';

import Nav from './common/Nav';

type BreadcrumbNavLink = {
  url: string;
  text: string;
};

export type BreadcrumbNavigation = {
  parentLinks: BreadcrumbNavLink[];
  currentPage: BreadcrumbNavLink;
};

// These props only require a Cart if showNav is true. Discriminated unions FTW.
type Props =
  | {
      breadcrumbNav: BreadcrumbNavigation;
      showNav?: false;
    }
  | {
      breadcrumbNav: BreadcrumbNavigation;
      showNav: true;
      cart: DeathCertificateCart;
    };

const PageLayout: React.StatelessComponent<Props> = props => {
  const breadcrumbNavAttributes = {
    className: 'p-a300',
    ...props.breadcrumbNav,
  };

  return (
    <AppLayout
      secondaryNav={props.showNav && <Nav cart={props.cart} />}
      breadcrumbNav={<BreadcrumbNav {...breadcrumbNavAttributes} />}
    >
      {props.children}
    </AppLayout>
  );
};

export default PageLayout;
