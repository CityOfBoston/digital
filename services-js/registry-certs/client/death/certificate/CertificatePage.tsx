import React from 'react';
import { action } from 'mobx';
import { observer } from 'mobx-react';
import Head from 'next/head';
import Link from 'next/link';
import Router from 'next/router';

import { getDependencies, ClientContext, ClientDependencies } from '../../app';
import { DeathCertificate } from '../../types';

import Cart from '../../store/Cart';
import SiteAnalytics from '../../lib/SiteAnalytics';

import {
  CERTIFICATE_COST_STRING,
  PERCENTAGE_CC_STRING,
  FIXED_CC_STRING,
  SERVICE_FEE_URI,
} from '../../../lib/costs';

import AppLayout from '../../AppLayout';

interface DefaultProps {
  cart: Cart;
  siteAnalytics: SiteAnalytics;
}

interface InitialProps {
  id: string;
  certificate: DeathCertificate | null;
  backUrl: string | null;
}

interface Props extends InitialProps, Partial<DefaultProps> {}

interface State {
  quantity: number | null;
}

@observer
class CertificatePage extends React.Component<Props & DefaultProps, State> {
  state: State;
  quantityField: HTMLInputElement | null = null;

  static get defaultProps(): DefaultProps {
    const { cart, siteAnalytics } = getDependencies();
    return { cart, siteAnalytics };
  }

  static async getInitialProps(
    ctx: ClientContext,
    dependenciesForTest?: ClientDependencies
  ): Promise<InitialProps> {
    const {
      query: { id, backUrl },
      res,
    } = ctx;
    const { deathCertificatesDao } =
      dependenciesForTest || getDependencies(ctx);

    if (!id) {
      throw new Error('Missing id');
    }

    const certificate = await deathCertificatesDao.get(id);

    if (!certificate && res) {
      res.statusCode = 404;
    }

    return {
      id,
      certificate,
      backUrl,
    };
  }

  constructor(props: Props & DefaultProps) {
    super(props);

    const { id, cart } = props;

    this.state = {
      quantity: cart.getQuantity(id) || 1,
    };
  }

  componentWillMount() {
    const { siteAnalytics, id } = this.props;

    siteAnalytics.addProduct(id, 'Death certificate', 'Death certificate');
    siteAnalytics.setProductAction('detail');
  }

  setCartQuantity = action(
    'CertificatePageController setCartQuantity',
    async (quantity: number) => {
      const { certificate, cart, siteAnalytics } = this.props;

      if (certificate) {
        if (quantity === 0) {
          cart.remove(certificate.id);

          siteAnalytics.sendEvent('UX', 'click', 'add to cart');
        } else {
          cart.setQuantity(certificate, quantity);
          siteAnalytics.sendEvent('UX', 'click', 'add to cart');

          await Router.push('/death/cart');
          window.scroll(0, 0);
        }
      }
    }
  );

  handleQuantityChange = (
    ev: React.ChangeEvent<HTMLSelectElement | HTMLInputElement>
  ) => {
    const { value } = ev.target;

    if (value === 'other') {
      this.setState({ quantity: null });
      if (this.quantityField) {
        this.quantityField.focus();
      }
    } else {
      this.setState({
        quantity: value ? Math.max(0, Math.min(parseInt(value, 10), 99)) : null,
      });
    }
  };

  handleAddToCart = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault();

    const { quantity } = this.state;

    if (typeof quantity !== 'number') {
      return;
    }

    this.setCartQuantity(quantity);
  };

  setQuantityField = (quantityField: HTMLInputElement | null) => {
    this.quantityField = quantityField;
  };

  render() {
    const { id, certificate, backUrl } = this.props;

    const { firstName, lastName } = certificate || {
      firstName: null,
      lastName: null,
    };

    const fullName = certificate
      ? `${firstName || ''} ${lastName || ''}`
      : null;

    return (
      <AppLayout showNav>
        <div className="b-ff">
          <Head>
            <title>
              Boston.gov — Death Certificates — {fullName || `#${id}`}
            </title>
          </Head>

          <div className="b-c b-c--nbp b-ff">
            <div className="sh sh--b0">
              <h1 className="sh-title">
                {fullName || 'Certificate not found'}
              </h1>
            </div>

            {certificate &&
              certificate.pending && (
                <div className="br br--r br-a200 m-v300 p-a300 t--info">
                  This certificate is <strong>pending</strong> and will not
                  include the cause of death. Some insurance and banking
                  companies won’t accept a death certificate if it is still
                  pending.
                </div>
              )}

            <div className="m-v300 b-ff">
              {certificate && this.renderCertificate(certificate)}
              {!certificate && (
                <div className="t--info">
                  We could not find a certificate with ID #{id}.
                </div>
              )}
            </div>

            <div className="g g--r g--vc">
              <div className="g--5 m-v300">
                {certificate && this.renderAddToCart()}
              </div>

              <div className="g--7 m-v300">
                {backUrl ? (
                  <Link href={backUrl}>
                    <a style={{ fontStyle: 'italic' }}>
                      ← Back to search results
                    </a>
                  </Link>
                ) : (
                  <Link href="/death">
                    <a style={{ fontStyle: 'italic' }}>← Go to search</a>
                  </Link>
                )}
              </div>
            </div>
          </div>

          <div className="b--g m-t700">
            <div className="b-c b-c--smv t--subinfo">
              Death certificates cost {CERTIFICATE_COST_STRING} each. That price
              includes shipping. You will be charged an extra service fee of not
              more than {FIXED_CC_STRING} plus {PERCENTAGE_CC_STRING}. That fee
              goes directly to a third party to pay for the cost of card
              processing. Learn more about{' '}
              <a href={SERVICE_FEE_URI}>card service fees</a> at the City of
              Boston.
            </div>
          </div>
        </div>
      </AppLayout>
    );
  }

  renderCertificate({
    firstName,
    lastName,
    age,
    deathDate,
    deathYear,
    birthDate,
  }: DeathCertificate) {
    return (
      <div className="certificate">
        <ul className="dl">
          <li className="dl-i">
            <span className="dl-t">First name</span>
            <span className="dl-d">{firstName}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Last name</span>
            <span className="dl-d">{lastName}</span>
          </li>
          {birthDate && (
            <li className="dl-i">
              <span className="dl-t">Date of birth</span>
              <span className="dl-d">{birthDate}</span>
            </li>
          )}
          <li className="dl-i">
            <span className="dl-t">Date of death</span>
            <span className="dl-d">{deathDate || deathYear}</span>
          </li>
          {age && (
            <li className="dl-i">
              <span className="dl-t">Age</span>
              <span className="dl-d">{age}</span>
            </li>
          )}
        </ul>

        <style jsx>{`
          .certificate {
            display: flex;
            flex-direction: column;
            flex-grow: 1;
            justify-content: space-between;
          }
          .dl-i {
            display: flex;
            align-items: center;
            justify-content: flex-start;
          }
          .dl-t {
            padding-right: 1em;
            width: 30%;
          }
          .dl-t,
          .dl-d {
            line-height: 1rem;
            vertical-align: center;
          }
        `}</style>
      </div>
    );
  }

  renderAddToCart() {
    const { cart, id } = this.props;
    const { quantity } = this.state;

    const cartQuantity = cart.getQuantity(id);

    return (
      <form onSubmit={this.handleAddToCart} className="js-add-to-cart-form">
        <div className="m-r100">
          <input
            ref={this.setQuantityField}
            type="text"
            name="quantity"
            aria-label="Quantity"
            className="txt-f txt-f--combo txt-f--auto ta-r"
            size={3}
            value={
              typeof quantity === 'number' && quantity >= 0 ? quantity : ''
            }
            onChange={this.handleQuantityChange}
          />
          <div className="sel-c sel-c--sq quantity-dropdown">
            <select
              name="quantityMenu"
              aria-label="Quantity menu"
              value={
                typeof quantity === 'number' && quantity <= 10
                  ? quantity
                  : 'other'
              }
              className="sel-f sel-f--sq"
              onChange={this.handleQuantityChange}
            >
              {cartQuantity
                ? [
                    <option value="0" key="remove">
                      Remove
                    </option>,
                    <option disabled key="separator">
                      ---------------
                    </option>,
                  ]
                : null}
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
              <option value="5">5</option>
              <option value="6">6</option>
              <option value="7">7</option>
              <option value="8">8</option>
              <option value="9">9</option>
              <option value="10">10</option>
              <option disabled>---------------</option>
              <option value="other">Other…</option>
            </select>
          </div>
        </div>
        <button
          type="submit"
          className="btn btn--row"
          disabled={quantity === cartQuantity || quantity === null}
        >
          {cartQuantity ? 'Update Cart' : 'Add to Cart'}
        </button>
        <style jsx>{`
          form {
            display: flex;
            align-items: center;
          }

          .quantity-dropdown:after {
            content: 'Qty.';
          }
        `}</style>
      </form>
    );
  }
}

export default (CertificatePage as any) as React.ComponentClass<Props> & {
  getInitialProps: (typeof CertificatePage)['getInitialProps'];
};
