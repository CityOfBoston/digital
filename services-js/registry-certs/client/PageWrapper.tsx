/** @jsx jsx */

import { css, jsx } from '@emotion/core';

import { Component, ReactNode } from 'react';

import {
  ProgressBar,
  // MEDIA_LARGE,
  // MEDIA_SMALL,
  MEDIA_X_LARGE,
} from '@cityofboston/react-fleet';

import PageLayout from './PageLayout';

import { BREADCRUMB_NAV_LINKS } from '../lib/breadcrumbs';

type CertificateRequestType = 'birth' | 'marriage';

export interface Progress {
  totalSteps: number;
  currentStep: number;
  currentStepCompleted: boolean;
}

interface Props {
  certificateType: CertificateRequestType;
  progress?: Progress;
  footer?: ReactNode;
}

/**
 * Convenience component for freestanding pages.
 *
 * Used for Birth and Marriage certificate requests.
 */
export default class PageWrapper extends Component<Props> {
  render() {
    const { certificateType, progress, footer, children } = this.props;

    return (
      <PageLayout breadcrumbNav={BREADCRUMB_NAV_LINKS[certificateType]}>
        <div
          className={`b-c b-c--hsm ${footer ? 'b-c--nbp' : ''}`}
          aria-live="polite"
        >
          <h1 className="sh-title" css={titleStyle(certificateType)}>
            Request a {certificateType} certificate
          </h1>

          {progress && <ProgressBar {...progress} />}

          <section className="m-t500">{children}</section>
        </div>

        {footer}
      </PageLayout>
    );
  }
}

// We put a hard top limit on the size, since with the responsive sizes on
// sh-title the phrase “Request a birth certificate” gets too big and wraps
// past 1245px wide. Since “marriage” contains more characters, the max font
// size is dropped to 34px.
function titleStyle(type: CertificateRequestType) {
  const styling = {
    paddingRight: 0,
    [MEDIA_X_LARGE]: {
      fontSize: type === 'birth' ? 38 : 34,
    },
  };

  // todo: adjust sizing further for Marriage
  // const isMarriage =
  //   type === 'marriage'
  //     ? { fontSize: 16, [MEDIA_SMALL]: { fontSize: 22 }, [MEDIA_LARGE]: { fontSize: 25 } }
  //     : null;

  return css({
    // ...isMarriage,
    ...styling,
  });
}
