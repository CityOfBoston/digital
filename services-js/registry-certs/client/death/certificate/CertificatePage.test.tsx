import React from 'react';
import { mount } from 'enzyme';

import { GaSiteAnalytics } from '@cityofboston/next-client-common';

import DeathCertificateCart from '../../store/DeathCertificateCart';
import DeathCertificatesDao from '../../dao/DeathCertificatesDao';

import CertificatePage from './CertificatePage';

import { TYPICAL_CERTIFICATE } from '../../../fixtures/client/death-certificates';

jest.mock('next/router');
jest.mock('../../dao/DeathCertificatesDao');

describe('getInitialProps', () => {
  let deathCertificatesDao;

  beforeEach(() => {
    deathCertificatesDao = new DeathCertificatesDao(null as any);
  });

  it('loads the cert passed in query', async () => {
    deathCertificatesDao.get.mockReturnValue(TYPICAL_CERTIFICATE);

    const initialProps = await CertificatePage.getInitialProps(
      { query: { id: '000002' }, res: undefined },
      { deathCertificatesDao }
    );

    expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
    expect(initialProps).toMatchSnapshot();
  });

  it('handles a 404', async () => {
    deathCertificatesDao.get.mockReturnValue(null);

    const initialProps = await CertificatePage.getInitialProps(
      { query: { id: '000002' }, res: undefined },
      { deathCertificatesDao }
    );

    expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
    expect(initialProps).toMatchSnapshot();
  });

  describe('operations', () => {
    let scrollSpy;

    beforeEach(() => {
      scrollSpy = jest.spyOn(window, 'scroll').mockImplementation(() => {});
    });

    afterEach(() => {
      scrollSpy.mockRestore();
    });

    describe('setCartQuantity', () => {
      let cart;
      let siteAnalytics;
      let component;

      beforeEach(() => {
        cart = new DeathCertificateCart();
        siteAnalytics = new GaSiteAnalytics();

        component = mount(
          <CertificatePage
            deathCertificateCart={cart}
            siteAnalytics={siteAnalytics}
            id="0002"
            certificate={TYPICAL_CERTIFICATE}
            backUrl="/search?q=jayne"
          />
        ).instance();
      });

      it('adds 5 things to the cart', () => {
        component.setCartQuantity(5);
        expect(cart.size).toEqual(5);
      });

      it('removes certificates when set to 0', () => {
        cart.setQuantity(TYPICAL_CERTIFICATE, 5);
        expect(cart.entries.length).toBe(1);

        component.setCartQuantity(0);
        expect(cart.entries.length).toBe(0);
      });
    });
  });
});
