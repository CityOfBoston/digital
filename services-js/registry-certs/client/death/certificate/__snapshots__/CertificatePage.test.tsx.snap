// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInitialProps handles a 404 1`] = `
Object {
  "backUrl": null,
  "certificate": null,
  "id": "000002",
}
`;

exports[`getInitialProps loads the cert passed in query 1`] = `
Object {
  "backUrl": null,
  "certificate": Object {
    "age": "53",
    "birthDate": "5/1/1962",
    "deathDate": "3/4/2016",
    "deathYear": "2016",
    "firstName": "BRUCE",
    "id": "000002",
    "lastName": "BANNER",
    "pending": false,
  },
  "id": "000002",
}
`;
